### PR TITLE
[Slightly less WIP] Design SystemParams to hold state

### PR DESCRIPTION
**Pull request by DJMcNab**
⚠ Mad scientist mode engaged ⚠

My goal for this design work is to allow:

```rust
fn my_system(Local<u32>, Query<Entity>, Res<T>, Local<Vec<u32>>);
```
to be added with the locals as 
```rust
let a = vec![10];
[..].add_system(my_system.system().configure(move |c| {c.0 = Some(15); c.3 = Some(a)}))
```
as opposed to the current
```rust
let sys = my_system.system();
app.insert_local(sys.id(), 15u32);
let a = vec![10u32];
app.insert_local(sys.id(), a);
[..].add_system(sys);
```

Hopefully I won't destroy a mid-sized city along the way.

### CODE REVIEW

**Comment by DJMcNab** [comment]
Discord discussion: https://discord.com/channels/691052431525675048/749335865876021248/805195040368361502

**Comment by cart** [comment]
This seems really nice if we can get it working. I'm a little bit curious about how you are going to persist this configuration inside a boxed system though. That feels like the hard part of the problem (and why we use a non-generic SystemState right now).

**Comment by DJMcNab** [comment]
Ideally, we'll avoid persisting the config and only persist the state, but yeah I see your point

We'll just have to see how it goes

**Comment by DJMcNab** [comment]
This looks like it might work, so I need some bikeshedding from someone about the locals syntax.
Option 1:
```rust
fn uses_locals(i_use_this_as_input: &Local<Vec<T>>, i_edit_this: &mut Local<u32>){}
```
Option 2:
```rust
fn uses_locals(i_use_this_as_input: Local<Vec<T>>, mut i_edit_this: Local<u32>){}
```
In terms of signature, I prefer option 1. But the problem is that it takes two derefs to get to T (such as for mutation). If anyone has a clever plan, let me know.

**Comment by cart** [comment]
I like option (2). Local is a "smart pointer". Passing in a reference to a "smart pointer" feels redundant and introduces boilerplate. (see Res, ResMut, Mut, etc)

**Comment by DJMcNab** [comment]
In this PR, for option 1 `Local` would be exactly `struct Local<T>(T)`, i.e. not a smart pointer, just a local value.

Option 3 would be:
```rust
fn uses_locals(i_use_this_as_input: &Vec<T>, i_edit_this: &mut u32){}
```

**Comment by cart** [comment]
Ahh gotcha gotcha. That does make things more interesting. I retract my previous statement and I now don't have any strong opinions yet. `&Local<T>` does feel too verbose though. Would the new interface allow us to store a `T` in the param config, but return a `Local<T>` smart pointer to that value?

**Comment by DJMcNab** [comment]
The beauty is that we don't need a smart pointer at all

We have a mutable reference to the local, so we can just give that to the system.

File: crates/bevy_ecs/src/system/mod.rs [hunk 1]
```diff
@@ -14,3 +14,7 @@ pub use query::*;
 pub use system::*;
 pub use system_chaining::*;
 pub use system_param::*;
+
+// Electric boogaloo
```
Comment (cart) [inline crates/bevy_ecs/src/system/mod.rs:18]: :laughing:

**Comment by cart** [comment]
Given how much the system construction api has changed, it might be worth trying to remove `.system()` in the add_system api just to see if it works. Although (1) that should probably wait until after queries / resources are re-added and (2) we don't need to overload the scope of this pr. We should only try it here if its easy.

**Comment by DJMcNab** [comment]
Right, I've started work on the fallout.

It's not complete yet, but I think it's enough that with another hour or two of work it will be ready. In particular, I need to update the derives to work with this new style (creating an equivalent `Config` and `State` types)

In particular, now is the best chance (this time) to comment on the `Local` api. I'm leaning towards `Local<T>` being effectively `Local<T>(&mut T)`, but with the deref being impled - that is, the same UX as present. Also I have just realised that option 3 collides with `Commands`, so that's probably out for now

File: crates/bevy_ecs/src/system/mod.rs (hunk 1 above)
Comment (DJMcNab) [reply crates/bevy_ecs/src/system/mod.rs]: It was easier to manage things creating two concurrent versions than to modify the existing one in-place.

Comment (cart) [reply crates/bevy_ecs/src/system/mod.rs]: Makes sense to me. I just like fun comments like this :)

**Comment by cart** [comment]
> In particular, now is the best chance (this time) to comment on the Local api. I'm leaning towards Local<T> being effectively Local<T>(&mut T), but with the deref being impled - that is, the same UX as present

Yeah I think thats the right call.

> Also I have just realised that option 3 collides with Commands, so that's probably out for now

Yeah makes sense to me. We can always revisit later. I have a feeling that `&mut Commands` will go back to being `mut Commands` in the bevy ecs core rewrite. We need to remove the entity reserver api and the best / simplest way I can think of is to do this:

```rust
Commands<'a> {
   world: &'a World,
   inner: &'a mut InnerCommands,
}
```

**Comment by DJMcNab** [comment]
And done :tm:

**Comment by DJMcNab** [comment]
I want to combine a few highly related changes once #1144 lands. So this is likely going to need a full rewrite.

But I'm also saying this would be ready to merge if you felt like doing so.

**Comment by cart** [comment]
Yesterday I noticed that there are a lot of synergies between the work here and my ecs rewrite (largely related to system query state). Specifically, it will allow us to migrate to a typed interface for cached query state and store it inside the system. This will also allow me to rework the "query fetch" interface to be more "stateful" as well. I decided to experiment a bit and have successfully incorporated the following SystemParam interface (which builds and runs as expected):

```rust
pub trait SystemParam: Sized {
    type State: SystemParamState;
    type Fetch: for<'a> FetchSystemParam<'a, State = Self::State>;
}

pub trait SystemParamState: Send + Sync + 'static {
    /// runs once to retrieve the initial value, which is then stored in the system 
    fn init(world: &mut World) -> Self;
    
    /// runs once per frame (for updating params according to things like new Archetypes)
    #[inline]
    fn update(&mut self, world: &World, system_state: &mut SystemState) {}
    
    /// runs once at the end of a stage. for doing things like applying Commands 
    #[inline]
    fn apply(&mut self, _world: &mut World) {}
}

pub trait FetchSystemParam<'a> {
    type Item;
    type State: 'a;
    unsafe fn get_param(
        state: &'a mut Self::State,
        system_state: &'a SystemState,
        world: &'a World,
    ) -> Option<Self::Item>;
}
```

Full system.rs: https://gist.github.com/cart/d585e0512c8dc97d1a56f363cfc5ead4
Full system_param.rs: https://gist.github.com/cart/529d8f91ea0f0937061604429e9ee770

I think the interface above works pretty well: the code is now _tiny_. As you mentioned, SystemState had a lot of trimdown potential.

This also worked without any new abstractions like AsSystem. It all just integrates with the current approach.

The only major missing piece is that it currently doesn't support user-provided overrides of param state (which I know is one of the major selling points of the approach in the pr). Given that the config lives inside structs in FuncSystem, I think this is pretty trivially accomplished by adding a new trait, which we could impl for the relevant types (FuncSystem, InputFuncSystem):

```rust
// names TBD
trait ParamStateBuilder<Params> {
  fn config_params(self, func: impl FnOnce(params: &mut Params)) -> Self;
}
```

Which would then enable

```rust
app.add_system(
  foo.system().config_params(|params| params.0 = something)
)
```

Another nuance: in my impl systems store param state in an `Option<Params>` and they are all initialized together. We would need to modify the storage to be more like `(Option<P1>, Option<P2>, ...)` to support the interface you want in this pr. However it would be nice to avoid unwrapping Options every frame, so we might want to use something unsafe like MaybeUninit instead (and somehow enforce initialization safety in the config_params interface. maybe with tracking bools stored separately).

Any thoughts on the proposed interface? Any major downsides I'm missing?

Lets try to align on an interface, make the whatever changes we need to here, merge this pr before my core rewrite pr (so you get proper credit), and I can adapt that pr as needed.

**Comment by DJMcNab** [comment]
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=29415f5e9c27b3465030c3aaf7e0205b is basically my next plan for this, with the headline feature being:
 ```rust
 fn main() {
    fn my_system(first: FirstThree) { dbg!(first); }
    accepts_system(my_system);
    accepts_system(|it: LastThree| {
        dbg!(it);
    });
    let x = vec![1, 2, 3, 4];
    accepts_system(move |it: LastThree| {
        dbg!(it);
        dbg!(&x);
    });
}
 ```

**Comment by alice-i-cecile** [comment]
Is there a way to serialize `Local` resources? Especially as part of `Scenes`. I'm a bit nervous about expanding this pattern without having a plan to take that into account, although admittedly this PR looks like it might provide nice hooks for loading saved values back in.

**Comment by DJMcNab** [comment]
I don't know how that could possibly work without a stable way to refer to systems?

Like the correct way to think about that is that local resources should be transient, such as for EventReaders and the like

If the Local contains important state, then I think we'd have a problem.

**Comment by alice-i-cecile** [comment]
> I don't know how that could possibly work without a stable way to refer to systems?
> 
> Like the correct way to think about that is that local resources should be transient, such as for EventReaders and the like
> 
> If the Local contains important state, then I think we'd have a problem.

Good, I can get behind that philosophy, although it should be clearly documented. This came up in discussions about storing reproducible RNGs in `Local` variables, which, if we can't serialize them, is definitely the wrong approach.

**Comment by alice-i-cecile** [comment]
@cart this has been obsoleted by / incorporated into #1525 right?

**Comment by cart** [comment]
Largely, yes. This also adds some new system construction trait stuff that enables a couple of things:
1. Removes the `Option<SystemParam>` from function systems in favor of a "config type" -> "system type" lifecycle
2. Removes the need for `.system()`

There are tradeoffs (largely just more complexity / types related to system construction), but its worth considering. However at this point I think its better to create a new pr for that.

**Pull request closed without merging**